### ISSUE TITLE
Provide easy way to mock modules, and then un-mock them

### ISSUE THREAD

**Issue by internettrans** [CanopyTax/node-jspm-jasmine#32]
System.delete System.set System.get, etc is the main approach for this I think. the systemjs-hotloader is also another good place to see how to delete modules and then reload them. Some of the tricks with this is that we need all files who have already imported a mocked module to get the new module when we mock it.

**Comment by internettrans** [comment]
I'm thinking maybe something like a global function `mockModule('name-of-module', { default: 1, func1: function() {}, value3: '32444243' })` would be a good api. And with a corresponding `unmockModule('name-of-module')` that could be called in the `afterEach`, in case people don't want to mock a specific module in all of their tests.

**Comment by m-a-r-c-e-l-i-n-o** [comment]
@joeldenning I'm not totally convinced that this should be embedded. Taking [mock-require](https://github.com/boblauer/mock-require) and [mockery](https://github.com/mfncooper/mockery) as precedence, I made [this](https://github.com/m-a-r-c-e-l-i-n-o/jspm-mock), which is meant to go inside your specs. Let me know if that resolves the issue at hand, otherwise, we can grab the code from there (if useful), modify, and embed. I'm still working out some kinks, and it doesn't work in latest (beta) versions of jspm do to a problem with `System.define`, but it's being looked at.

**Comment by internettrans** [comment]
Your project looks awesome -- I haven't tried it out because all of my projects are using jspm@beta and you said that that isn't working. What issue with `System.define` are you referring to? Have you looked into `System.set('fs', new Module({val1: '1'}))`

**Comment by m-a-r-c-e-l-i-n-o** [comment]
@joeldenning Very cool, I'll let you know when it's ready for `jspm@beta`. [This](https://github.com/jspm/jspm-cli/issues/1863) is the issue with `System.define`, it's pending release, though. I did try `System.set`, but the problem is that for it to work, the module being set needs to be in the registry. As far as I know, the only ways to get a module into the registry is through `system.import` or `system.define`. I opted for `System.define` to make the mock method synchronous, thus reducing complexity for the user. Do you know of a better strategy?

**Comment by internettrans** [comment]
Nice, maybe I'll just pull down the master branch of es6-module-loader and install it locally into projects to try out `jspm-mock`. A question about a use case with jspm-mock, though: if module `./mod1.js` is loaded with a dependency `./mod2.js`, and then you mock `./mod2.js`, and unmock `./mod2.js`, which copy of `mod2` will `mod1` be referencing at that point? Sorry if that's confusing. Another way of putting it is that I'm not sure if `System.delete` is sufficient to ensure that all modules get the old version of a mocked dependency. See https://github.com/m-a-r-c-e-l-i-n-o/jspm-mock/blob/master/src/jspm-mock.js#L57

**Comment by m-a-r-c-e-l-i-n-o** [comment]
@joeldenning So here's how I can explain what would happen (assuming I accurately understood what you're saying)...

```
var mod1 = jspmMock.get('./mod1.js') // with a dependency called "./mod2.js"
jspmMock.mock('./mod2.js', 'fake-module-of-sorts') // variable "mod1" will still be referencing the old "./mod2.js".
```

For you to get "mod1.js" to take the newly mocked "mod2.js" into account, you would need to do:
`var mod1 = jspmMock.get('./mod1.js') // again`

I don't think that there is anything that could be done about that because the references are in the variables, if they were maybe stored in object properties inside jspm-mock (i.e. `jspmMock.mod1`) then we could change this behavior. I would argue that this is likely not necessary, do you have a strong need for this kind of behavior? How do you think it could be improved? The `systemjs-hotloader` module you mentioned above reloads all modules whose dependency is affected, but that only works because the main scripts are reloaded also, thus returning to its beginning state, which is not an issue for its intended use cases (although I believe that there are ways to maintain some state with react). Hard to reason how that could be applied for these unit testing use cases since you can't easily just pause and resume the script execution.

**Comment by internettrans** [comment]
@m-a-r-c-e-l-i-n-o The description in #33 is sort of a continuation of the conversation here.

Also, I think that the issue we were just discussing about what "unmocking" really does and how it impacts modules that already imported things, it's something that I'm definitely interested in exploring. In some of my tests, being able to mock/unmock with all deps being updated would be really powerful. For example, in my test for `foo` (which imports `bar`) I might want to mock `bar`. But then I might also want to write a test for bar, in which case I don't want bar to be mocked anymore.

**Comment by internettrans** [comment]
Fixed with #33, except for the unmocking part
